yolox/utils/logger.py

```diff
@@ -438,3 +438,329 @@ def initialize_wandb_logger(cls, args, exp, val_dataset):
                     wandb_params.update({k[len(prefix):]: v})
 
         return cls(config=vars(exp), val_dataset=val_dataset, **wandb_params)
+
+import importlib.metadata
+import importlib.util
+from collections.abc import MutableMapping
+import packaging.version
+import json, platform, time
+
+ENV_VARS_TRUE_VALUES = {"1", "ON", "YES", "TRUE"}
+
+
+
+class MlflowLogger(object):
+    """
+    Logging training runs with hyperparamter, datasets and trained models to MLflow.
+    Mlflow support Model Tracking, Experiment Tracking, and Model Registry.
+    It can be hosted on-premises or in all the major cloud provider or with databricks also.
+
+    For changing default logging Behaviour you can change mlflow environment variables:
+        https://mlflow.org/docs/latest/python_api/mlflow.environment_variables.html
+
+    For more information, please refer to:
+    https://mlflow.org/docs/latest/introduction/index.html
+    """
+    def __init__(self):
+        if not self.is_required_library_available():
+            raise RuntimeError("MLflow Logging requires mlflow and python-dotenv to be installed. Run `pip install mlflow python-dotenv`." )
+
+        import mlflow
+        from dotenv import find_dotenv, load_dotenv
+        load_dotenv(find_dotenv())
+        self._MAX_PARAM_VAL_LENGTH = mlflow.utils.validation.MAX_PARAM_VAL_LENGTH
+        self._MAX_PARAMS_TAGS_PER_BATCH = mlflow.utils.validation.MAX_PARAMS_TAGS_PER_BATCH
+        self._initialized = False
+        self._auto_end_run = False
+        self.best_ckpt_upload_pending = False
+        self._ml_flow = mlflow
+
+    def is_required_library_available(self):
+        "check if required libraries are available."
+        dotenv_availaible = importlib.util.find_spec("dotenv") is not None
+        mlflow_available = importlib.util.find_spec("mlflow") is not None
+        return dotenv_availaible and mlflow_available
+
+    def flatten_dict(self, d: MutableMapping, parent_key: str = "", delimiter: str = "."):
+        """Flatten a nested dict into a single level dict."""
+
+        def _flatten_dict(d, parent_key="", delimiter="."):
+            for k, v in d.items():
+                key = str(parent_key) + delimiter + str(k) if parent_key else k
+                if v and isinstance(v, MutableMapping):
+                    yield from self.flatten_dict(v, key, delimiter=delimiter).items()
+                else:
+                    yield key, v
+
+        return dict(_flatten_dict(d, parent_key, delimiter))
+
+    def setup(self, trainer):
```

Comment:
I notice that only trainer.exp and trainer.args is needed here. Should we use `exp, args` instead of `trainer`?

UPDATE: If you make sure that you might use some attributes in trainer in the future, please use `trainer` as args. However, if you make sure that you will not use more attrs in the future, some specific arguments is better.
This princple holds for all the following methods.

Reply:
removed trainer as input from all function and further chunked it into exact dependency.